trieste/acquisition/rule.py

```diff
@@ -425,36 +391,39 @@ def acquire(
 
         y_min = tf.reduce_min(dataset.observations, axis=0)
 
-        if state is None:
-            eps = 0.5 * (global_upper - global_lower) / (5.0 ** (1.0 / global_lower.shape[-1]))
-            is_global = True
-        else:
-            tr_volume = tf.reduce_prod(
-                state.acquisition_space.upper - state.acquisition_space.lower
-            )
-            step_is_success = y_min < state.y_min - self._kappa * tr_volume
-
-            eps = (
-                state.eps
-                if state.is_global
-                else state.eps / self._beta
-                if step_is_success
-                else state.eps * self._beta
-            )
+        def go(state: TrustRegion.State | None) -> tuple[TrustRegion.State | None, TensorType]:
+
+            if state is None:
+                eps = 0.5 * (global_upper - global_lower) / (5.0 ** (1.0 / global_lower.shape[-1]))
+                is_global = True
+            else:
+                tr_volume = tf.reduce_prod(
+                    state.acquisition_space.upper - state.acquisition_space.lower
+                )
+                step_is_success = y_min < state.y_min - self._kappa * tr_volume
+
+                eps = (
+                    state.eps
+                    if state.is_global
+                    else state.eps / self._beta
+                    if step_is_success
+                    else state.eps * self._beta
+                )
 
-            is_global = step_is_success or not state.is_global
+                is_global = step_is_success or not state.is_global
 
-        if is_global:
-            acquisition_space = search_space
-        else:
-            xmin = dataset.query_points[tf.argmin(dataset.observations)[0], :]
-            acquisition_space = Box(
-                tf.reduce_max([global_lower, xmin - eps], axis=0),
-                tf.reduce_min([global_upper, xmin + eps], axis=0),
-            )
+            if is_global:
+                acquisition_space = search_space
+            else:
+                xmin = dataset.query_points[tf.argmin(dataset.observations)[0], :]
+                acquisition_space = Box(
+                    tf.reduce_max([global_lower, xmin - eps], axis=0),
+                    tf.reduce_min([global_upper, xmin + eps], axis=0),
+                )
 
-        acquisition_function = self._builder.prepare_acquisition_function(datasets, models)
-        point = self._optimizer(acquisition_space, acquisition_function)
-        state_ = TrustRegion.State(acquisition_space, eps, y_min, is_global)
+            points = self._rule.acquire(acquisition_space, datasets, models)
+            state_ = TrustRegion.State(acquisition_space, eps, y_min, is_global)
+
+            return state_, points
 
-        return point, state_
+        return go
```

Comment:
Why does acquire now return a function?

Reply:
it's how i've made the `state` part of the return type rather than a built-in part of acquisition rules.

an alternative approach could be to simply make `TrustRegion` have some mutable state. There are two reasons I didn't do this. One is historical: trieste was like this when I joined the team. The other is because it's usually good form to avoid mutable state

Reply:
it's perhaps better to think of it as `TrustRegion` returns a `State`ful set of points. That is, the value of the points depending on some state, and computing them updates that state. That it's a function is merely a choice of representation for state transitions